### DISCUSSION TITLE
Strange behavior when CapsLock is On

### DISCUSSION THREAD

**Discussion by k1mk-7388** [rustdesk/rustdesk#9060] [General]
Hi everyone, initial question: when I try using RustDesk (Windows to Windows)  and have CapLocks On on the client side some (but not all) of the ALT+alpha keys do not seem to be received by the remote side. Alt+W does work, but Alt+Q and Alt+U don't seem to be getting across. Any ideas where I should look first?

**Comment by rustdesk** [comment]
@fufesou note this, have a look after finishing web client.

**Comment by fufesou** [comment]
@k1mk-7388 Hi, please give more infos. I'll try fix this issue in the future.

1. Which keyboard mode do you use?

<img width="272" alt="1723644278244" src="https://github.com/user-attachments/assets/736b7fac-6b9d-4a79-998f-beee5228d17c">


2. What're the input language on both sides? eg: en-US -> fr-FR

3. `client side` is the local (controlling) side or remote (controlled) side?

4. How does TeamViewer or other software behave in your case?

**Reply by k1mk-7388** [reply]
3 client side is local ; remote side is the system to which local is connecting 

4  We don't have TeamViewer installed but AnyDesk does work as expected in this case

**Comment by k1mk-7388** [comment]
1. I am using Legacy Mode but it's behaving the same way in Map Mode and
Translate mode beta

2. en-US -> en-US

**Comment by rustdesk** [comment]
More capslock issue here, https://github.com/rustdesk/rustdesk/issues/8760, and https://github.com/rustdesk/rustdesk/issues/9617

**Comment by fufesou** [comment]
@k1mk-7388 Hi, what app do you used to test ?

Maybe the key simulations are not work as expected.

I've record the keys on the remote side (win -> win).

### Keyboard mode keys

I use the attached app to grab keys. [rustdesk-keys-grab.zip](https://github.com/user-attachments/files/17355279/rustdesk-keys-grab.zip)


When CapsLock is On, pressing `Alt` + `<alpha>` including `Alt` + `U` and `Alt` + `Q`, keys are sent as follows.

Take `Alt` + `Q` as the example.

#### map mode

```
alt down

CapsLock down
CapsLock up
q down
CapsLock down
CapsLock up

alt up

CapsLock down
CapsLock up
q up
CapsLock down
CapsLock up
```

#### translate mode beta

```
alt down
q down
alt up
q up

```

### web keyboard test

You can confirm the keys here
https://www.keyboardtester.com/tester.html

**Comment by k1mk-7388** [comment]
The application that I am testing with is N1MM Logger Plus <
https://www.n1mm.com>,  a Windows native application which is written in VB
IIRC

Since my original inquiry, the versions of RustDesk on both the server and
client sides  of the connection I am testing over have been updated to the
latest public release.  I suspect that I did not retest with any of the
modes other than legacy mode after updating -- legacy mode still doesn't
work.

Now, upon re-testing both Map mode and Translate mode mode beta do "work"
-- "work" as in the remote application responds as  expected in the case(s)
that I identified although neither mode achieves this result with the
unadorned simplicity of either local keyboard entry or how the AnyDesk
remote desktop application (or the Microsoft Remote Desktop application for
that matter)  perform the identical operation (see full testing results
below).

Legacy mode seems to be missing the release (KeyUp) for the Q key -- that
may be a reason that mode is failing. There also seems to be an irrelevant
release for the KannaMode key (with no corresponding press) present as well
as a press and no release for the  LeftShift key. Not too surprising that
the remote side might have some difficulty in deciphering what all that was
intended to indicate.

It is not at all obvious exactly what the extra overheads of all the
superfluous traffic in the non-parsimonious cases of legacy and map modes
are intended to achieve. Are there defined use cases for those three modes?
If I might ask, where can I find further documentation on each of the
Legacy, Map and Translate modes and the circumstance under which those
modes are intended to be used? User documentation covering those modes is
scant to non-existent. Just a pointer to the relevant modules in the source
tree would probably be sufficient.

Based on this latest testing, it would appear that Translate mode beta may
be what comes closest to matching  the functional behavior of other remote
desktop apps. But even then, that mode still seems to consistently emit a
redundant key code on release which is undesirable as it constitutes an
unnecessary overhead.

Hope this is useful feedback

Thanks,
Michael

Full testing results for Alt-Q follows

*For local keyboard:*

*CapsLock On*

name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*CapsLock On*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("Q"), unicode: [81], is_dead: false }),
type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038

*Using AnyDesk (another remote desktop application which is known to work
correctly with this application) to connect*

*CapsLock Off*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010
name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*CapsLock On*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("Q"), unicode: [81], is_dead: false }),
type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*Using RustDesk *

*Caps Lock Off*

*Client Legacy mode*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038

name: None, type: KeyRelease(KanaMode), code: 0x00, scan: 0x0000


*Client Map mode*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*Client Translate mode beta*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*Caps Lock On*


*Client Legacy mode:*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyPress(ShiftLeft), code: 0xA0, scan: 0x002A

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(KanaMode), code: 0x00, scan: 0x0000

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038


*Client Map mode*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: Some(UnicodeInfo { name: Some("Q"), unicode: [81], is_dead: false }),
type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyPress(CapsLock), code: 0x14, scan: 0x003A

name: None, type: KeyRelease(CapsLock), code: 0x14, scan: 0x003A


*Client Translate mode beta:*


name: None, type: KeyPress(Alt), code: 0xA4, scan: 0x0038

name: Some(UnicodeInfo { name: Some("q"), unicode: [113], is_dead: false
}), type: KeyPress(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(KeyQ), code: 0x51, scan: 0x0010

name: None, type: KeyRelease(Alt), code: 0xA4, scan: 0x0038

**Comment by fufesou** [comment]
Thanks for your detailed feedback.

https://github.com/rustdesk/rustdesk/wiki/FAQ#keyboard-translation-modes

You should not use the "Legacy mode".

> But even then, that mode still seems to consistently emit a redundant key code on release which is undesirable as it constitutes an unnecessary overhead.

Yes, there're extra `KeyRelease` and `CapsLock` on the controlled side. 
Some redundant keys (`CapsLock` in "Map mode") can indeed be removed. We may change this.
The extra `KeyRelease` is to ensure that the key will not be released. Such a key is worth the extra operation because it does not cost much and will ensure stability.
